internal/fs/dir_handle.go

```diff
@@ -156,78 +182,76 @@ func fixConflictingNames(entries []fuseutil.Dirent) (err error) {
-	// like inode IDs based on (object name, generation) hashes. But then what
-	// about the birthday problem? And more importantly, what about our
-	// semantic of not minting a new inode ID when the generation changes due
-	// to a local action?
-	for i, _ := range entries {
-		entries[i].Inode = fuseops.RootInodeID + 1
-	}
+// Fetch Dirent entries from GCSfuse.Will be used as a goroutine which is run asynchronously
+// to fetch data in the background while kernel requests are also served simultaneously.
+func (dh *DirHandle) FetchEntriesAsync(
+	rootInodeId int,
+	firstCall bool) {
+
+	//New context is needed as the parent goroutine exiting earlier than the child will cause the
+	//context to be cancelled prematurely
+	ctx := context.Background()
+	var err error
+	var entryForSorting fuseutil.Dirent
+	//ContinuationToken is also empty in case of firstCall and after all entries have been fetched.
+	//Keeping continuation token local so as to lessen the time for which the mutex is held
+	//Keep fetching entries in batches of MaxResultsForListObjectsCall
+	var ContinuationToken string
+	for ContinuationToken != "" || firstCall {
+		var entries []fuseutil.Dirent
+		dh.In.Lock()
+		entries, ContinuationToken, err = dh.In.ReadEntries(ctx, ContinuationToken)
+		dh.In.Unlock()
+		if err != nil {
+			err = fmt.Errorf("ReadEntries: %w", err)
+			dh.Mu.Lock()
+			dh.err = err
+			dh.Mu.Unlock()
+			//Signal the suspended go routine that an error has occurred.
+			dh.cond.Broadcast()
+			break
+		}
+		//Use the last entry from the last fetch for fixing naming conflicts
+		if !firstCall {
+			entries = append(entries, entryForSorting)
+		}
+		sort.Sort(sortedDirents(entries))
+		err = fixConflictingNames(entries)
 
-	return
-}
+		if err != nil {
+			err = fmt.Errorf("fixConflictingNames: %w", err)
+			dh.Mu.Lock()
+			dh.err = err
+			dh.Mu.Unlock()
+			//Signal the suspended go routine that an error has occurred.
+			dh.cond.Broadcast()
+			break
+		}
 
-// LOCKS_REQUIRED(dh.Mu)
-// LOCKS_EXCLUDED(dh.in)
-func (dh *dirHandle) ensureEntries(ctx context.Context) (err error) {
-	dh.in.Lock()
-	defer dh.in.Unlock()
-
-	// Read entries.
-	var entries []fuseutil.Dirent
-	entries, err = readAllEntries(ctx, dh.in)
-	if err != nil {
-		err = fmt.Errorf("readAllEntries: %w", err)
-		return
-	}
+		//Save the last entry from current fetch to use it for
+		//fixing naming conflicts for next fetch
+		if ContinuationToken != "" {
+			entryForSorting = entries[len(entries)-1]
+			entries = entries[:len(entries)-1]
+		}
+		dh.Mu.Lock()
+		//Update InodeID and Offset for the entries
+		for i := range entries {
+			entries[i].Inode = fuseops.InodeID(rootInodeId + 1)
+			entries[i].Offset = fuseops.DirOffset(uint64(len(dh.entries) + i + 1))
+		}
 
-	// Update state.
-	// Fix up offset fields.
-	for i := 0; i < len(entries); i++ {
-		entries[i].Offset = fuseops.DirOffset(uint64(len(dh.entries) + i + 1))
+		dh.entries = append(dh.entries, entries...)
+		dh.entriesValid = true
```

Comment:
do we really need this entriesValid property. can we live without it?

Reply:
Since the condition of starting fetch can be changed to 0 entries + fetchOver=false, this property can be dropped.

Reply:
entriesValid property is required in checkInvariants to detect inconsistency.